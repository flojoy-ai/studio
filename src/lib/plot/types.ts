export interface Drawable {
  draw: () => void;
}
