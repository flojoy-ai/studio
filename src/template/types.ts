export interface SomeObj {
    id: string,
    propertyX: string;
    propertyY: string;
}