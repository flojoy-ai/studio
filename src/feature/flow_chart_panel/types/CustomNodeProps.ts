export type ElementsData = {
  id: string;
  label: string;
  func: string;
  type: string;
  running?: boolean;
  failed?:boolean;
  ctrls: {
    [key: string]: {
      functionName: string;
      param: string;
      value: number;
    };
  };
  inputs?: Array<{ name: string; id: string; type:string }>
  selects?: any;
}

export interface CustomNodeProps {
  data: ElementsData;
}