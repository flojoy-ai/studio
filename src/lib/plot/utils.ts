export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
