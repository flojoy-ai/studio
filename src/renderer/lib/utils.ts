import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
