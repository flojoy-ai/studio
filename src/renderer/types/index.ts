export type {
  ALL_DC_TYPE,
  DataContainer,
  DataContainerType,
} from "./dataContainer";
export type { CustomNodeProps, BlockData } from "./node";
export type {
  DataContainer2PlotlyProps,
  OverridePlotData,
  PlotProps,
} from "./plotly";
