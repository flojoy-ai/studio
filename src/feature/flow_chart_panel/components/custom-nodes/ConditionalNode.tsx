import HandleComponent from "@feature/flow_chart_panel/components/HandleComponent";
import { CustomNodeProps } from "@feature/flow_chart_panel/types/CustomNodeProps";
import { useFlowChartState } from "@hooks/useFlowChartState";
import { Box, Text, clsx } from "@mantine/core";
import { useSocket } from "@src/hooks/useSocket";
import { memo, useEffect, useState } from "react";
import { useNodeStyles } from "../DefaultNode";
import NodeWrapper from "../NodeWrapper";
import NodeEditButtons from "../node-edit-menu/NodeEditButtons";

const ConditionalNode = ({ data }: CustomNodeProps) => {
  const { classes } = useNodeStyles();
  const [additionalInfo, setAdditionalInfo] = useState({});

  const { runningNode, failedNode } = useFlowChartState();
  const params = data.inputs || [];

  const { states } = useSocket();
  const { programResults } = states!;

  const isLoopInfoExist = () => {
    const isExist = Object.keys(additionalInfo).find(
      (value, _) => value === data.id
    );
    return isExist && data.func === "LOOP";
  };

  const current_iteration = isLoopInfoExist()
    ? additionalInfo[data.id]["current_iteration"] || 0
    : 0;
  const total_iteration = isLoopInfoExist()
    ? data["ctrls"][`LOOP_${data.label}_iteration_count`]["value"] || 0
    : 0;

  useEffect(() => {
    console.log("18");
    if (programResults?.io?.length! > 0) {
      let programAdditionalInfo = {};

      const results = programResults?.io;
      results?.forEach((element) => {
        programAdditionalInfo = {
          ...programAdditionalInfo,
          [element.id]: element["additional_info"],
        };
      });

      setAdditionalInfo(programAdditionalInfo);
    }
  }, [programResults]);

  return (
    <NodeWrapper data={data}>
      <Box
        className={clsx(
          runningNode === data.id || data.selected ? classes.defaultShadow : "",
          failedNode === data.id ? classes.failShadow : ""
        )}
      >
        <Box
          className={clsx(classes.nodeContainer, classes.defaultNode)}
          style={{
            ...(params.length > 0 && { padding: "0px 0px 8px 0px" }),
          }}
        >
          <Box>
            <Box mt={25}>{data.label}</Box>
            <Box>
              {data.func === "CONDITIONAL" && (
                <>
                  {params?.length !== 0 ? (
                    <Text
                      mt={20}
                      sx={{ textAlign: "center" }}
                      data-testid="conditional-operator-type"
                    >
                      x {data["ctrls"]["operator_type"]["value"]} y
                    </Text>
                  ) : (
                    <>
                      {Object.keys(additionalInfo)
                        .filter((value) => value === data.id)
                        .map((_, index) => {
                          return (
                            <Text key={index + 1}>
                              status: {additionalInfo[data.id]["status"]}
                            </Text>
                          );
                        })}
                    </>
                  )}
                </>
              )}
              {data.func === "TIMER" && (
                <Text data-testid="timer-value">
                  {data["ctrls"][`TIMER_${data.label}_sleep_time`]["value"]}s
                </Text>
              )}
              {data.func === "LOOP" && (
                <Box data-testid="loop-info">
                  <p>{`${current_iteration}/${total_iteration}`}</p>
                </Box>
              )}
            </Box>
          </Box>

          <Box
            display="flex"
            h={params.length > 0 ? (params.length + 1) * 32 : "fit-content"}
            sx={{
              flexDirection: "column",
            }}
          >
            <HandleComponent data={data} inputs={params} />
          </Box>
        </Box>
      </Box>
    </NodeWrapper>
  );
};

export default memo(ConditionalNode);
