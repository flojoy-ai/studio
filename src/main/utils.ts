import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
