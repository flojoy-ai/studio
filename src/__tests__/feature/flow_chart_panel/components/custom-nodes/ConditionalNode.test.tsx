import { CustomNodeProps } from "@src/feature/flow_chart_panel/types/CustomNodeProps";
import ConditionalNode from "@src/feature/flow_chart_panel/components/custom-nodes/ConditionalNode";
import { renderWithTheme } from "@src/__tests__/__utils__/utils";

const handleRemove = jest.fn();

const props: CustomNodeProps = {
  data: {
    id: "test-id",
    label: "test",
    func: "test",
    type: "test",
    ctrls: {},
    path: "",
    inputs: [
      {
        id: "test",
        name: "test",
        type: "test",
      },
    ],
  },
  handleRemove,
};

jest.mock("@hooks/useFlowChartState");

jest.mock("@feature/flow_chart_panel/components/HandleComponent", () => {
  const mockChildren = jest
    .fn()
    .mockReturnValue(<div data-testid="handle-component" />);
  return { __esModule: true, default: mockChildren };
});

jest.mock("@src/hooks/useSocket");

describe("ConditionalNode", () => {
  it("checks the snapshot", () => {
    const { container } = renderWithTheme(<ConditionalNode {...props} />);
    expect(container).toMatchSnapshot();
  });
  it.each([
    ["CONDITIONAL", "conditional-operator-type"],
    ["TIMER", "timer-value"],
    ["LOOP", "loop-info"],
  ])("checks if the component: %p is rendered", (componentName, testId) => {
    let ctrls = {};
    if (componentName === "CONDITIONAL") {
      ctrls = {
        operator_type: {
          value: ">",
        },
      };
    } else if (componentName === "TIMER") {
      ctrls = {
        TIMER_test_sleep_time: {
          value: 2,
        },
      };
    }

    const { getByTestId } = renderWithTheme(
      <ConditionalNode
        handleRemove={handleRemove}
        data={{ ...props.data, func: componentName, ctrls: ctrls }}
      />
    );

    const component = getByTestId(testId);
    expect(component).toBeInTheDocument();
  });
});
