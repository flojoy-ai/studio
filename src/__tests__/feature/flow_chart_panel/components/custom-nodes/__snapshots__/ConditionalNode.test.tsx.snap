// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionalNode checks the snapshot 1`] = `
<div>
  <div
    data-testid="node-wrapper"
  >
    <div
      class="mantine-1avyp1d"
    >
      <div
        class="mantine-1oexgir"
        style="padding: 0px 0px 8px 0px;"
      >
        <div
          class="mantine-1avyp1d"
        >
          <div
            class="mantine-11k3zi0"
          >
            test
          </div>
          <div
            class="mantine-1avyp1d"
          />
        </div>
        <div
          class="mantine-m3buv7"
        >
          <div
            data-testid="handle-component"
          />
        </div>
      </div>
    </div>
  </div>
  ,
</div>
`;
